70_FLIPFluids/main.cpp

```diff
@@ -23,2218 +18,1845 @@ using namespace asset;
-			memBarrier.srcAccessMask = ACCESS_FLAGS::SHADER_WRITE_BITS;
-			memBarrier.dstStageMask = PIPELINE_STAGE_FLAGS::COMPUTE_SHADER_BIT;
-			memBarrier.dstAccessMask = ACCESS_FLAGS::SHADER_READ_BITS;
-			cmdbuf->pipelineBarrier(E_DEPENDENCY_FLAGS::EDF_NONE, {.memBarriers = {&memBarrier, 1}});
+        if (m_surface)
+            return { { m_surface->getSurface() } };
+
+        return {};
+    }
+
+    inline bool onAppInitialized(smart_refctd_ptr<ISystem>&& system) override
+    {
+        m_inputSystem = make_smart_refctd_ptr<InputSystem>(logger_opt_smart_ptr(smart_refctd_ptr(m_logger)));
+
+        if (!device_base_t::onAppInitialized(std::move(system)))
+            return false;
+        if (!asset_base_t::onAppInitialized(std::move(system)))
+            return false;
+
+        // init grid params
+        usePreset(CENTER_DROP);
+        
+        WorkgroupCountParticles = (numParticles + WorkgroupSize - 1) / WorkgroupSize;
+        WorkgroupCountGrid = {
+            (m_gridData.gridSize.x + WorkgroupGridDim - 1) / WorkgroupGridDim,
+            (m_gridData.gridSize.y + WorkgroupGridDim - 1) / WorkgroupGridDim,
+            (m_gridData.gridSize.z + WorkgroupGridDim - 1) / WorkgroupGridDim
+        };
+
+        {
+            float zNear = 0.1f, zFar = 10000.f;
+            core::vectorSIMDf cameraPosition(14, 8, 12);
+            core::vectorSIMDf cameraTarget(0, 0, 0);
+            matrix4SIMD projectionMatrix = matrix4SIMD::buildProjectionMatrixPerspectiveFovLH(core::radians(60.0f), float(WIN_WIDTH) / WIN_HEIGHT, zNear, zFar);
+            camera = Camera(cameraPosition, cameraTarget, projectionMatrix, 1.069f, 0.4f);
+
+            m_pRenderParams.zNear = zNear;
+            m_pRenderParams.zFar = zFar;
+        }
+        m_pRenderParams.radius = m_gridData.gridCellSize * 0.4f;
+
+        // create buffers
+        video::IGPUBuffer::SCreationParams params = {};
+        params.size = sizeof(SGridData);
+        params.usage = IGPUBuffer::EUF_UNIFORM_BUFFER_BIT | IGPUBuffer::EUF_TRANSFER_DST_BIT | IGPUBuffer::EUF_INLINE_UPDATE_VIA_CMDBUF;
+        createBuffer(gridDataBuffer, params);
+
+        params.size = 2 * sizeof(float32_t4);
+        createBuffer(pressureParamsBuffer, params);
+        
+        params.size = sizeof(SMVPParams);
+        createBuffer(cameraBuffer, params);
+
+        params.size = sizeof(SParticleRenderParams);
+        createBuffer(pParamsBuffer, params);
+
+        params.size = numParticles * sizeof(float32_t3);
+        params.usage = IGPUBuffer::EUF_STORAGE_BUFFER_BIT | IGPUBuffer::EUF_SHADER_DEVICE_ADDRESS_BIT;;
+        createBuffer(particleData.positionBuffer, params, IDeviceMemoryAllocation::EMAF_DEVICE_ADDRESS_BIT);
+        createBuffer(particleData.velocityBuffer, params, IDeviceMemoryAllocation::EMAF_DEVICE_ADDRESS_BIT);
+
+        params.size = numParticles * 6 * sizeof(VertexInfo);
+        createBuffer(particleVertexBuffer, params, IDeviceMemoryAllocation::EMAF_DEVICE_ADDRESS_BIT);
```

Comment:
you already have a position and velocity buffer above, why do you need a 6x AoS vertex buffer?

The whole point of programmable pulling is to that you can ready from BDA of the above using `gl_VertexIndex/4`